<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Подключение-датафрейма-и-библиотек" data-toc-modified-id="Подключение-датафрейма-и-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подключение датафрейма и библиотек</a></span><ul class="toc-item"><li><span><a href="#Установка-библиотек" data-toc-modified-id="Установка-библиотек-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Установка библиотек</a></span></li><li><span><a href="#Подключение-библиотек" data-toc-modified-id="Подключение-библиотек-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Подключение библиотек</a></span></li><li><span><a href="#Подключение-датафрейма" data-toc-modified-id="Подключение-датафрейма-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Подключение датафрейма</a></span></li></ul></li><li><span><a href="#Исследование-данных" data-toc-modified-id="Исследование-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Исследование данных</a></span><ul class="toc-item"><li><span><a href="#Пример-датафрейма" data-toc-modified-id="Пример-датафрейма-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Пример датафрейма</a></span></li><li><span><a href="#Информация-о-датафрейме" data-toc-modified-id="Информация-о-датафрейме-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Информация о датафрейме</a></span></li><li><span><a href="#Информация-о-числовых-признаках" data-toc-modified-id="Информация-о-числовых-признаках-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>Информация о числовых признаках</a></span></li><li><span><a href="#Информация-о-целевом-признаке" data-toc-modified-id="Информация-о-целевом-признаке-1.2.4"><span class="toc-item-num">1.2.4&nbsp;&nbsp;</span>Информация о целевом признаке</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-неинформативных-столбцов" data-toc-modified-id="Удаление-неинформативных-столбцов-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Удаление неинформативных столбцов</a></span></li><li><span><a href="#Изменение-типов--данных-столбцов" data-toc-modified-id="Изменение-типов--данных-столбцов-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Изменение типов  данных столбцов</a></span></li><li><span><a href="#Разбивка-на-обучающую-и-тестовую-выборки" data-toc-modified-id="Разбивка-на-обучающую-и-тестовую-выборки-1.3.3"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>Разбивка на обучающую и тестовую выборки</a></span></li></ul></li><li><span><a href="#Обработка-текстов" data-toc-modified-id="Обработка-текстов-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Обработка текстов</a></span><ul class="toc-item"><li><span><a href="#Очистка-текста" data-toc-modified-id="Очистка-текста-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Очистка текста</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-1.4.2"><span class="toc-item-num">1.4.2&nbsp;&nbsp;</span>Лемматизация</a></span></li></ul></li><li><span><a href="#Создание-признаков" data-toc-modified-id="Создание-признаков-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Создание признаков</a></span><ul class="toc-item"><li><span><a href="#Создание-N-грамм" data-toc-modified-id="Создание-N-грамм-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Создание N-грамм</a></span></li><li><span><a href="#TF-IDF" data-toc-modified-id="TF-IDF-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>TF-IDF</a></span></li></ul></li><li><span><a href="#Обьединение-всех-этапов-обработки-данных" data-toc-modified-id="Обьединение-всех-этапов-обработки-данных-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Обьединение всех этапов обработки данных</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Тестирование" data-toc-modified-id="Тестирование-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Тестирование</a></span><ul class="toc-item"><li><span><a href="#Проверка-модели-на-адекватность" data-toc-modified-id="Проверка-модели-на-адекватность-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Проверка модели на адекватность</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

### Подключение датафрейма и библиотек

#### Установка библиотек

In [ ]:
!pip install -q spacy
!pip install nltk
!pip install lightgbm
!pip install tqdm

In [1]:
!spacy download -q en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 684.4 kB/s eta 0:00:0000:0100:01
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


#### Подключение библиотек

In [2]:
from tqdm.auto import tqdm
import pandas as pd
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.preprocessing import FunctionTransformer
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from re import sub 
import numpy as np
from sklearn.model_selection import train_test_split
from nltk import download
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn.model_selection import StratifiedKFold
from sklearn.dummy import DummyClassifier
from copy import deepcopy
from lightgbm import LGBMClassifier
import warnings, spacy, sys, os

#### Подключение датафрейма

In [3]:
try:
    df = pd.read_csv('/datasets/toxic_comments.csv')
except FileNotFoundError:
    df = pd.read_csv('//Users//georgespesivtsev//Documents//toxic_comments.csv')

### Исследование данных

#### Пример датафрейма

In [4]:
df.head()

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

Столбец '<font color='green'>Unnamed: 0</font>' это столбец повторяющий индексы, его необходимо удалить.

#### Информация о датафрейме

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


Пропусков нет.

#### Информация о числовых признаках

In [6]:
df.describe()

Unnamed: 0          toxic
count  159292.000000  159292.000000
mean    79725.697242       0.101612
std     46028.837471       0.302139
min         0.000000       0.000000
25%     39872.750000       0.000000
50%     79721.500000       0.000000
75%    119573.250000       0.000000
max    159450.000000       1.000000

Аномальных значений в целевом признаке нет, тип данных избыточен, можно заменить тип '<font color='blue'>int64</font>' на '<font color='blue'>int8</font>'.

#### Информация о целевом признаке

In [7]:
df.toxic.value_counts()

0    143106
1     16186
Name: toxic, dtype: int64

В данных присутствует существенный дисбаланс классов.

### Предобработка данных

#### Удаление неинформативных столбцов

In [8]:
df = df.drop('Unnamed: 0', axis=1)

#### Изменение типов  данных столбцов

In [9]:
df.toxic = df.toxic.astype('int8')

#### Разбивка на обучающую и тестовую выборки

In [10]:
RANDOM_STATE = 41626
X_train, X_test, y_train, y_test = train_test_split(df.drop('toxic', axis=1), df.toxic, \
                                                    test_size=0.25, \
                                                    random_state=RANDOM_STATE, \
                                                    stratify=df.toxic)

### Обработка текстов

#### Очистка текста

In [11]:
cleaning = lambda t: ' '.join(sub(r"[^a-z' ]", '', t).split())

In [1]:
processing = Pipeline([(
    'clearing',
    FunctionTransformer( 
        lambda data: data.text.str.replace('\n', ' ').str.lower().apply(cleaning)
    )
)])

NameError: name 'Pipeline' is not defined

#### Лемматизация

In [13]:
nlp = spacy.load('en_core_web_sm', disable=['parser', 'ner'])
lemmatize = lambda t: " ".join([token.lemma_ for token in nlp(t) \
                                if not token.is_stop and \
                                token.lemma_ not in ['-PRON-', '"', "'", 'utc']])

In [14]:
colours = ['#ab97dc', '#c6f9d7', '#ffffd4', '#f7c6cd']
tqdm.pandas(desc="Лемматизация текстов: ", colour=colours.pop(0))
processing.steps.append((
    'lemmatize',
    FunctionTransformer(
        lambda data: data.progress_apply(lemmatize).values
    )
))

### Создание признаков

#### Создание N-грамм

In [15]:
download('stopwords')
stopwords = list(nltk_stopwords.words('english'))

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/georgespesivtsev/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [16]:
making_features = FeatureUnion([
 ('making_beg', CountVectorizer(stop_words=stopwords))   
], n_jobs=-1)

#### TF-IDF

In [17]:
making_features.transformer_list.append((
    'making_tf_idf',
    TfidfVectorizer(stop_words=stopwords)
))

### Обьединение всех этапов обработки данных

In [18]:
preprocessing = Pipeline([
    ('processing', processing),
    ('making_features', making_features)
])

## Обучение

Поиск лучшей модели и гиперпарамтров:

In [19]:
class best_model_search():
    def __init__(self, estimators, params_grid, preprocessing, cv=3):
        self.estimators = estimators if type(estimators) == list else [estimators]
        self.params_grid = params_grid
        self.preprocessing = preprocessing
        self.cv = cv
    
        
    def fit(self, X, y, research=False):
        
        if hasattr(self, 'model') and not research:
            self.model.fit(X, y)
            return self
        
        X = self.preprocessing.fit_transform(X)
        y = y.reset_index(drop=True)
        
        get_model_name = lambda model: str(type(model)).split('.')[-1][:-2]
        
        ests_results = [] 
        
        for est in tqdm(self.estimators, position=0, leave=True, desc='Обработанные модели:', colour=colours.pop(0)): 
            
            tqdm.write(get_model_name(est) + ':')
            
            est_results = [] 
            
            est_params = {}
            for param in self.params_grid.keys():
                if param in est.get_params().keys():
                    est_params[param] = self.params_grid[param]
            
            params_sets = [[]]
            
            for idx in range(len(est_params)): 
                params_sets = [list(x)+[y] for x in params_sets for y in list(est_params.values())[idx]]
            params_dicts = [{list(est_params.keys())[idx]: params_set[idx] for idx in range(len(params_set))} \
                            for params_set in params_sets]
            for params in tqdm(params_dicts, position=0, leave=True, desc='Обработанные наборы параметров:', colour=colours.pop(0)):
                
                f1_cv = []
                
                copy_est = deepcopy(est).set_params(**params)
                
                for train_index, valid_index in StratifiedKFold(n_splits=self.cv).split(X, y):
                    X_train_cv, y_train_cv = X[train_index, :], y.iloc[train_index]
                    X_valid, y_valid = X[valid_index, :], y.iloc[valid_index]
                    
                    copy_est.fit(X_train_cv, y_train_cv)
                    f1_cv.append(f1_score(y_valid, copy_est.predict(X_valid)))
                
                result = {}
                result['params'] = params
                result['f1_score'] = pd.Series(f1_cv).mean()
                est_results.append(result)
            
            best_model_by_params = sorted(est_results,key=lambda d: d['f1_score'])[-1]
            best_model_by_params['model'] = est.set_params(**best_model_by_params['params'])
            ests_results.append(best_model_by_params)
            
        ests_results.sort(reverse=True, key=lambda d: d['f1_score'])
        
        self.result = {}
        self.result['model'] = [get_model_name(x['model']) for x in ests_results]
        self.result['best_f1_score'] = [x['f1_score'] for x in ests_results]
        for param in self.params_grid.keys():
                self.result[param] = [x['params'].get(param, np.nan) for x in ests_results]
        self.result = pd.DataFrame(self.result).set_index('model')
        display(self.result)
        
        self.best_est = ests_results[0]['model']
        self.best_est.fit(X, y)
        self.model = Pipeline([
            ('preprocessing', preprocessing),
            ('modelling', deepcopy(self.best_est))
        ])
        
        return self
    

    def predict(self, X):
        return self.model.predict(X)

Будем выбирать лучшую модель между LogisticRegression и LGBMClassifier (и лучшие гиперпараметры к ним ):

In [ ]:
model_search = best_model_search(
    preprocessing=preprocessing,
    estimators=[
        LogisticRegression(),
        LGBMClassifier()
    ],
    params_grid={
        'C': [.1, 1., 50., 1e2],
        'max_iter': [100, 500, 1000],
        'n_estimators': [100, 500, 1000],
        'learning_rate': [.1, .3, .5, .65],
        'random_state': [RANDOM_STATE],
        'n_jobs': [-1],
        'class_weight': ['balanced']
        
    },
    cv=X_train.shape[0] // X_test.shape[0]
)

Выключим предупреждения о том, что модели не сходятся при обучении. (работает только такой способ, из-за того, что предупреждения не перехватваются при использовании нескольких ядер, а также потому, что библиотеки numpy и sklearn переопределяют пространство предупреждений, даже после внесенных вами в него изменений)

In [21]:
if not sys.warnoptions:
    warnings.simplefilter('ignore')
    os.environ['PYTHONWARNINGS'] = 'ignore'

In [22]:
model_search.fit(X_train, y_train)

Лемматизация текстов:   0%|          | 0/119469 [00:00<?, ?it/s]

Обработанные модели::   0%|          | 0/2 [00:00<?, ?it/s]

LogisticRegression:


Обработанные наборы параметров::   0%|          | 0/12 [00:00<?, ?it/s]

LGBMClassifier:


Обработанные наборы параметров::   0%|          | 0/12 [00:00<?, ?it/s]

f1_score    C  max_iter  n_estimators  learning_rate  \
model                                                                      
LGBMClassifier      0.759646  NaN       NaN        1000.0            0.1   
LogisticRegression  0.758559  1.0    1000.0           NaN            NaN   

                    random_state  n_jobs class_weight  
model                                                  
LGBMClassifier             41626      -1     balanced  
LogisticRegression         41626      -1     balanced

Лучшей моделью оказалась LightGBM, её и будем использовать для предсказания тестовой выборки.

## Тестирование

In [23]:
y_pred = model_search.predict(X_test)

Лемматизация текстов:   0%|          | 0/39823 [00:00<?, ?it/s]

In [24]:
f1_score(y_test, y_pred)

0.7726423902894491

Метрика F1 больше 0.75, модель неплохо справляется со своей задачей.

### Проверка модели на адекватность

In [32]:
dummy_model = DummyClassifier(strategy='stratified')
dummy_model.fit(X_train, y_train)

DummyClassifier(strategy='stratified')

In [33]:
f1_score(y_test, dummy_model.predict(X_test))

0.10136986301369862

Подобранная модель значительно превоосходит константную модель, тем самым являясь адекватной.

## Общий вывод

- Были подключены нужные библиотеки и датафрейм с данными
- Данные исследованы:
    - Обнаружен столбец повторяющий индекс датафрейма
    - Пропуски и аномальные значения отсутствуют
    - В целевой метрике обнаружен дисбаланс классов
    - Целевой признак использует избыточный тип данных
- Данные предобработаны:
    - Удален столбцец повторяющий индекс датафрейма
    - Тип данных целевого признака изменен на минимально-достаточный
- Был создан алгоритм (pipeline) для предобработки данных, включающий в себя:
    - Обработку текстов:
        - Очистку от неподходящих символов
        - Лемматизацию
    - Создания признаков для моделлирования:
        - Создание униграмм
        - TF-IDF
- Данные разбиты на обучающую и тестовую выборки в соотношении 3:1
- Был создан алгоритм (class) использующий стратифицированную (из-за дисбаланса классов в целевом признаке) кросс-валидацию для поиска лучшей модели и гиперпараметров, из поданых ему на вход
- По результатам поиска лучшей моделью оказалась LightGBM с подобранными гиперпараметрами.
- На тествой выборке модель показала значение метрики F1 около 0.77
- Модель прошла проверку на адекватность, значительно превосходя константную модель